zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/internal/cache/ZiplineCache.kt

```diff
@@ -87,47 +99,68 @@ internal class ZiplineCache internal constructor(
   suspend fun getOrPut(
     applicationName: String,
     sha256: ByteString,
-    download: suspend () -> ByteString?
+    download: suspend () -> ByteString?,
   ): ByteString? {
     val read = read(sha256)
     if (read != null) return read
 
-    val contents = download() ?: return null
-    write(applicationName, sha256, contents)
-    return contents
+    val content = download() ?: return null
+    write(applicationName, sha256, content, isManifest = false)
+    return content
   }
 
-  fun read(
-    sha256: ByteString
-  ) = read(sha256.hex())
+  internal fun getOrPutManifest(
+    applicationName: String,
+    content: ByteString,
+  ): Files? {
+    val sha256 = content.sha256()
+    val metadata = getOrNull(sha256)
+    return metadata ?: write(applicationName, sha256, content, isManifest = true)
+  }
 
-  fun read(
-    sha256hex: String
-  ): ByteString? {
-    val metadata = database.filesQueries.get(sha256hex).executeAsOneOrNull() ?: return null
+  fun read(sha256: ByteString): ByteString? {
+    val metadata = database.filesQueries.get(sha256.hex()).executeAsOneOrNull() ?: return null
+    return read(metadata)
+  }
+
+  private fun read(metadata: Files): ByteString? {
     if (metadata.file_state != FileState.READY) return null
+
     // Update the used at timestamp
     database.filesQueries.update(
       id = metadata.id,
       file_state = metadata.file_state,
       size_bytes = metadata.size_bytes,
       last_used_at_epoch_ms = nowMs(),
     )
-    return try {
-      fileSystem.read(path(metadata)) {
+    val path = path(metadata)
+    val result = try {
+      fileSystem.read(path) {
         readByteString()
       }
     } catch (e: FileNotFoundException) {
-      null // Might have been pruned while we were trying to read.
+      null // Might have been pruned while we were trying to read?
+    }
+
+    if (result == null || result.sha256() != metadata.sha256_hex.decodeHex()) {
+      // File is absent or corrupt. Delete quietly.
+      try {
+        fileSystem.delete(path)
+        database.filesQueries.delete(metadata.id)
+      } catch (ignored: IOException) {
+      }
```

Comment:
nit: comment in the empty catch block for what case it is catching, I guess the parameter name works too